### ISSUE TITLE
Reading float value from tfrecord, return a not same value.

### ISSUE THREAD

**Issue by TechBK** [tensorflow/tensorflow#12876]
### System information
- OS Platform and Distribution: Linux Ubuntu 16.04)
- TensorFlow installed from: [binary](https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow_gpu-1.3.0-cp36-cp36m-linux_x86_64.whl)
- TensorFlow version: v1.3.0-rc2-20-g0787eee 1.3.0
- Python version: Python 3.6.2 :: Continuum Analytics, Inc.

### Describe the problem
When i read feature `label` with `float` type from tfrecord, it return not same value. Example:
- `3.3 -> 3.29999995`
- `3.7 -> 3.70000005`
- `4.9 -> 4.9000001`

Here, code make `Example`:
```
def make_example(s0, s1, label):
    ex = tf.train.SequenceExample()
    ex.context.feature["len_s0"].int64_list.value.append(len(s0))
    ex.context.feature["len_s1"].int64_list.value.append(len(s1))
    ex.context.feature["label"].float_list.value.append(label)# HERE: the feature are `float` type

    for w in s0:
        ex.feature_lists.feature_list["s0"].feature.add().int64_list.value.append(w)
    for w in s1:
        ex.feature_lists.feature_list["s1"].feature.add().int64_list.value.append(w)

    return ex
```
Parsing function:
```
def _parse_function(example_proto):
    context_features = {
        "len_s0": tf.FixedLenFeature((), dtype=tf.int64),
        "len_s1": tf.FixedLenFeature((), dtype=tf.int64),
        "label": tf.FixedLenFeature((), dtype=tf.float32) # HERE: the feature are `float` type
    }
    sequence_features = {
        "s0": tf.FixedLenSequenceFeature((), dtype=tf.int64),
        "s1": tf.FixedLenSequenceFeature((), dtype=tf.int64)
    }

    context_parsed, sequence_parsed = tf.parse_single_sequence_example(
        serialized=example_proto,
        context_features=context_features,
        sequence_features=sequence_features
    )

    len_s0 = tf.cast(context_parsed['len_s0'], dtype=tf.int8)
    len_s1 = tf.cast(context_parsed['len_s1'], dtype=tf.int8)
    # label = tf.convert_to_tensor(context_parsed['label'], dtype=tf.float32)
    # label = tf.round(context_parsed['label'])
    label = context_parsed['label']
    s0 = tf.cast(sequence_parsed['s0'], dtype=tf.int32)
    s1 = tf.cast(sequence_parsed['s1'], dtype=tf.int32)

    return {"len_s0": len_s0, "s0": s0, "len_s1": len_s1, "s1": s1}, label
```
When i call `result = tf.contrib.learn.run_n({"label": label}) `, it return:
```
{'label': array([ 3.29999995], dtype=float32)} # right value: 3.3
{'label': array([ 3.70000005], dtype=float32)} # right value: 3.7
{'label': array([ 4.9000001], dtype=float32)} # right value: 4.9
```

Some one can help me? Plzz told me some ways to fix it.

**Comment by quaeler** [comment]
This is a problem at the ProtocolBuffer layer. (@martinwicke i don't know who to tag about this)

If i do the following using your code:
```
def main(unused_argv):
	example = make_example([3, 5, 7], [4, 6, 8], 4.3)
	print example
	
	serialized = example.SerializeToString()

	reconstituted = tf.train.SequenceExample()
	reconstituted.ParseFromString(serialized)
	print reconstituted
	
	something, label = _parse_function(serialized)
	result = tf.contrib.learn.run_n({"label": label})
	print result
```

I'll get:
```
context {
  feature {
    key: "label"
    value {
      float_list {
        value: 4.3
      }
    }
  }
...
```
for `example`, and
```
context {
  feature {
    key: "label"
    value {
      float_list {
        value: 4.30000019073
      }
    }
  }
...
```
for `reconstituted` (and ultimately `[{'label': 4.3000002}]` for `result`)

I ran this against Python 2.7.10, and TF built from source from the head of master 5 days ago.

**Comment by martinwicke** [comment]
To me, this looks like a bug in protobuf. I'm wondering:

1. is this reproducible with a simpler proto, say a message containing only a single float?
2. is this reproducible using a language other than python? I.e., is the bug in the python binding, or deeper?

**Comment by quaeler** [comment]
Addressing (1.) first. Yes, this happens with a simple message:
```
(tensorflow_20170902) autogenic:12876 loki$ more simple_float.proto 
syntax = "proto3";

option cc_enable_arenas = true;
option java_outer_classname = "SimpleFloatProtos";
option java_multiple_files = true;
option java_package = "org.tensorflow.ldq.skratch";

package tensorflow;

message SimpleFloat {
        float value = 1;
}
(tensorflow_20170902) autogenic:12876 loki$ ~/arbeit/worg/projects/tensorflow/bazel-out/host/bin/external/protobuf_archive/protoc --python_out=. simple_float.proto 
(tensorflow_20170902) autogenic:12876 loki$ python
Python 2.7.10 (default, Feb  7 2017, 00:08:15) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import simple_float_pb2
>>> sf = simple_float_pb2.SimpleFloat()
>>> sf.value = 12.17
>>> print sf
value: 12.17

>>> serialized = sf.SerializeToString()
>>> reconstitution = simple_float_pb2.SimpleFloat()
>>> reconstitution.ParseFromString(serialized)
>>> print reconstitution
value: 12.1700000763

>>> quit()
(tensorflow_20170902) autogenic:12876 loki$ 
```
Will futz around for (2.) tonight or this weekend.

**Comment by quaeler** [comment]
Just to be replete, i have verified that (1.) occurs in the Python releases of TF 1.3.0 and 1.2.0

**Comment by martinwicke** [comment]
It's pretty clearly a protobuf issue. The question is why. Allegedly it's
only in the C++ backend. Another thing that could be tried.

Either way, a workaround is to avoid Protobuf.

**Comment by quaeler** [comment]
Agreed on both cases; i'm still ignorant enough of the codebase to know at what points (if any) protobuf is unavoidably used in the TF code and so could not practically be avoided by TF consumers.

Will still spend some time this weekend in probing and gedankenexperiment WRT (2.)

**Comment by martinwicke** [comment]
Actually, it's all much simpler. Python uses float64 internally (Python "float" is really double). Once you cast 4.3 to 4.3f you get 4.30000019073486.

Proof:
```
cat > foo.cpp <<EOF
#include <iostream>
#include <iomanip>
int main(){
  double a = 4.3;
  float b = (float) a;
  double c = (double) b;
  std::cout<<std::setprecision(15)<<a<<" "<<b<<" "<<c<<std::endl;
}
EOF
g++ foo.cpp && ./a.out
```
yields
```
4.3 4.30000019073486 4.30000019073486
```

Thanks @hawkinsp, @aselle!

Casting explicitly to `np.float32` and printing that before serialization should yield the same. The confusing part is that default print options will prevent you from seeing anything useful.

```python
>>> import numpy as np
>>> np.set_printoptions(precision=15)
>>> x = 4.3
>>> y = np.float32(x)
>>> print y
4.3
>>> y 
4.3000002
>>> x
4.3
```

**Comment by quaeler** [comment]
Interesting - thanks. ... but something is still rotten here; this isn't really TF's problem, but it seems like a real problem that will affect Python users of TF and so maybe that at least warrants some note in a read-me somewhere?

There's a use case problem which highlights the data representation problem.

The use case problem is: i should be able to do math operations on the values in my protobuf object prior to serialization, and have them be numerically equivalent to the result were i to perform those same operations on a deserialized version of that object.

The related data representation problem is: once it comes out of deserialization, that true original value isn't easily reclaimable (beyond some truly hacky schlock.)

Illustrated using the above simple float proto object & r1.3:

```python
>>> sf = simple_float_pb2.SimpleFloat()
>>> sf.value = 12.17
>>> cubedSF = math.pow(sf.value, 3)
>>> cubedSF
1802.485313
>>> npSF = np.float32(sf.value)
>>> npSF
12.17
>>> print npSF
12.17
>>> cubedNPSF = math.pow(npSF, 3)
>>> cubedNPSF  # now it is revealed that numpy was holding on to precision-schmutz
1802.4853468994372
>>> serialized = sf.SerializeToString()
>>> reconstitution = simple_float_pb2.SimpleFloat()
>>> reconstitution.ParseFromString(serialized)
>>> print reconstitution.value
12.1700000763
>>> reconstitution.value
12.170000076293945
>>> sf.value == reconstitution.value   # non-trivial problem
False
>>> np.float32(sf.value) == reconstitution.value  # but i could always write a custom comparator for pre & post serialization objects
True
>>> np.equal(sf.value, reconstitution.value)   # maybe using Numpy will solve everything
False
>>> cubedRe = math.pow(reconstitution.value, 3)
>>> cubedRe
1802.4853468994372
>>> npRe = np.float32(reconstitution.value)
>>> npRe   # actually unexpected - ?thank you Numpy?
12.17
>>> print npRe
12.17
>>> sf.value == npRe   # i can always dream
False
>>> np.equal(sf.value, npRe)  # really dream
False
>>> cubedNPRe = math.pow(npRe, 3)
>>> cubedNPRe
1802.4853468994372
>>> import StringIO  # were i intent on reclaiming the original value...
>>> hack = StringIO.StringIO()
>>> print >>hack, npRe
>>> hack.getvalue()
'12.17\n'
>>> originalValue = float(hack.getvalue())
>>> originalValue
12.17
>>> sf.value == originalValue
True
>>> np.equal(sf.value, originalValue)
True
>>> math.pow(originalValue, 3)
1802.485313
```
To get *consistent* behaviour, i should cast sf.value via np.float32() to do operations; then i will have consistent values from operations on pre and post serialization versions of the object - of course, those values would be wrong all along.
To get *correct* behaviour, i need to do a string parsing hack on my deserialized float prior to operations; then i will be able to have consistent results from operations on pre and post serialization versions of the object, and will have correct logic function results as well.

Am i cuckoo-banana to think this situation is a valid problem?

**Comment by Bengt** [comment]
I just ran into this, too. I wanted to make sure my importer code is correct, so I also implemented an exporter. Surprisingly, the results were not the same as the inputs. For my use case, the loss in precision that is described in this issue actually matters. My examples are drawn from a problem with a fractal result space. So deviating in the problem space by losing some precision actually makes my examples wrong, as @quaeler noted. This means that beyond 32 bit precision, my estimator would be learning garbage. However, I found that the work around of storing the data as strings and casting them back and forth is actually common.

Here is a blog post from the future (May 2019), encoding floats as bytes:

http://jrmeyer.github.io/machinelearning/2019/05/29/tensorflow-dataset-estimator-api.html

The money bit for serialization:

```python
def _float64_feature(float64_value):
    float64_bytes = [str(float64_value).encode()]
    bytes_list = tf.train.BytesList(value=float64_bytes)
    bytes_list_feature = tf.train.Feature(bytes_list=bytes_list)

    return bytes_list_feature
```

... and for deserialization:

```python
def deserialize(serialized_example):
    features = {
        'float_value': tf.FixedLenFeature([], tf.string),
    }
    features = tf.parse_single_example(
        serialized_example,
        features=features
    )

    features['float_value'] = tf.strings.to_number(string_tensor=features['float_value'], out_type=tf.float64)

    return features['float_value']
```

**Comment by dvaldivia** [comment]
@Bengt Did you ever found a solution around this without casting? I guess the protocol buffer looses all it's advantage of storing floats if we store them as strings :(

**Comment by dvaldivia** [comment]
I'm still seeing this issue with a simple serialization of a `tf.Tensor` (shape=(1, 512), dtype=float32) which is stored as a `list of floats` when the `TFRecord` example is created all the values are what is expected, but when reading the data back through the parse, the value of the `tf.Tensor` that is all wrong, all values are rounded up even if we are going from `float32` to `float32` .

**Comment by AbuAliAmin** [comment]
although the issue is close.. but i set this answer for my future reference.. I had this issue when writing audio files as floating point matrix using FloatList.. but when i used BytesList and stored the data into tfrecords and then read the data by decoding it.. the issue resolved.. note that even decoding with tf.float32 will lead not solve the issue. we need to decode it with tf.float64..

`def _bytes_feature2(value):
  """Returns a bytes_list from a string / byte."""
  return tf.train.Feature(bytes_list=tf.train.BytesList(value=[value]))

def serialize_example(sound):
  """
  Creates a tf.train.Example message ready to be written to a file.
  """
  # Create a dictionary mapping the feature name to the tf.train.Example-compatible
  # data type.
  feature = {
      'snd': _bytes_feature2(sound.tobytes()),
  }
  example_proto = tf.train.Example(features=tf.train.Features(feature=feature))
  return example_proto.SerializeToString()

def write_tfrecords(rf,snd):
    nsamples = len(snd)
    with tf.io.TFRecordWriter(rf) as writer:
        for i in range(nsamples):
            SND = snd[i]
            tf_example = serialize_example(SND)
            writer.write(tf_example)

# writing records
write_tfrecords(os.getcwd()+'\\tfrec\\'+'train.tfrecords',train)

# loading records
raw_dataset = tf.data.TFRecordDataset(os.getcwd()+'\\tfrec\\'+'train.tfrecords')

def parse_record(record):
  name_to_features= {
      'user_id':tf.io.FixedLenFeature([],tf.string),
      'utterance':tf.io.FixedLenFeature([],tf.string),
  }
  return tf.io.parse_single_example(record, name_to_features)

def decode_record(record):
  aud = tf.io.decode_raw(
    record['utterance'], out_type=tf.float64
  )
  usn = record['user_id']
  return aud  , usn

for record in raw_dataset:
  parsed_record = parse_record(record)
  decoded_record = decode_record(parsed_record)
  aud, usn = decoded_record
  print(aud.numpy()[0:10])
  print(train[0][0:10])


`
output:

[ 417.69951205 -231.58708746  -10.05624011 -146.10342256  -66.60317323
 -159.91550792   -3.93602823   29.94517981  106.22196629   65.53008959]
[ 417.69951205 -231.58708746  -10.05624011 -146.10342256  -66.60317323
 -159.91550792   -3.93602823   29.94517981  106.22196629   65.53008959]
